src/comm.cc

```diff
@@ -1579,6 +1579,12 @@ checkTimeouts(void)
 void
 commStartHalfClosedMonitor(int fd)
 {
+    if (!TheHalfClosed->empty()) {
+        debugs(5, 5, HERE << "half_closed is on fd = " << fd );
+        commCallCloseHandlers(fd);
+        commStopHalfClosedMonitor(fd);
+    }
+
     debugs(5, 5, "adding FD " << fd << " to " << *TheHalfClosed);
     assert(isOpen(fd) && !commHasHalfClosedMonitor(fd));
```

Comment:
If you want to work around this assertion without fixing the underlying bug, then the condition of the proposed new `if` statement would have to change to become specific to the condition(s) in this assertion. And the handling of that condition would have to be changed to printing an `Squid BUG 4156: ...` ERROR, probably followed by `return;` (but that needs to be tested -- I do not know enough about this bug to propose the exact workaround).

Do you know what assertion condition is failing, `isOpen(fd)` or `!commHasHalfClosedMonitor(fd)`? If you do not, you can find out by splitting that assert() into two, the way it should have been written:

```suggestion
    assert(isOpen(fd));
    assert(!commHasHalfClosedMonitor(fd));
```

One of the above assertions should fail in your tests (without the proposed `if` statement workaround!). Once we know which assertion fails (without the proposed `if` statement), we can start working on adjusting the proposed `if` statement accordingly.

Needless to say, it would be much better to actually fix the bug, but I do not have enough free time right now to triage and volunteer a proper fix, unfortunately.